sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowStringInterpolator.java

```diff
@@ -0,0 +1,192 @@
+  public static final String DD = "$DD";
+  private static final Set<String> WINDOWING_METADATA =
+      Sets.newHashSet(WINDOW, PANE_INDEX, YYYY, MM, DD);
+  private static final Pattern TEMPLATE_PATTERN = Pattern.compile("\\{(.+?)}");
+
+  /**
+   * @param template a String template, potentially with placeholders in the form of curly braces,
+   *     e.g. {@code "my {foo} template"}. During interpolation, these placeholders are replaced
+   *     with values in the Beam Row. For more details and examples, refer to the top-level
+   *     documentation.
+   * @param rowSchema {@link Row}s used for interpolation are expected to be compatible with this
+   *     {@link Schema}.
+   */
+  public RowStringInterpolator(String template, Schema rowSchema) {
+    this.template = template;
+
+    Matcher m = TEMPLATE_PATTERN.matcher(template);
+    fieldsToReplace = new HashSet<>();
+    while (m.find()) {
+      fieldsToReplace.add(StringUtils.strip(m.group(), "{}"));
+    }
+
+    List<String> rowFields =
+        fieldsToReplace.stream()
+            .filter(f -> !WINDOWING_METADATA.contains(f))
+            .collect(Collectors.toList());
+
+    RowFilter.validateSchemaContainsFields(rowSchema, rowFields, "string interpolation");
+  }
+
+  /** Performs string interpolation on the template using values from the input {@link Row}. */
+  public String interpolate(Row row) {
+    String interpolated = this.template;
+    for (String field : fieldsToReplace) {
+      List<String> levels = Splitter.on(".").splitToList(field);
+
+      Object val = MoreObjects.firstNonNull(getValue(row, levels, 0), "");
+
+      interpolated = interpolated.replace("{" + field + "}", String.valueOf(val));
+    }
+    return interpolated;
+  }
+
+  /** Like {@link #interpolate(Row)} but also potentially include windowing information. */
+  public String interpolate(Row row, BoundedWindow window, PaneInfo paneInfo, Instant timestamp) {
+    String interpolated = this.template;
+    for (String field : fieldsToReplace) {
+      Object val;
+      switch (field) {
+        case WINDOW:
+          val = window.toString();
+          break;
+        case PANE_INDEX:
+          val = paneInfo.getIndex();
+          break;
+        case YYYY:
+          val = timestamp.getChronology().year().get(timestamp.getMillis());
+          break;
+        case MM:
+          val = timestamp.getChronology().monthOfYear().get(timestamp.getMillis());
+          break;
+        case DD:
+          val = timestamp.getChronology().dayOfMonth().get(timestamp.getMillis());
+          break;
+        default:
+          List<String> levels = Splitter.on(".").splitToList(field);
+          val = MoreObjects.firstNonNull(getValue(row, levels, 0), "");
+          break;
+      }
+
+      interpolated = interpolated.replace("{" + field + "}", String.valueOf(val));
+    }
+    return interpolated;
+  }
+
+  private @Nullable Object getValue(@Nullable Row row, List<String> fieldLevels, int index) {
```

Comment:
Nit: Rather than recursing on a list of fieldLevels and an index, it might be simpler to do something like

```
def getValue(row, path):
  dot = path.indexOf('.')
  if dot == -1:
    return row.getValue(path)
  else:
    getValue(row.getValue(path[:dot]), path[:dot+1])
```

Reply:
Looks much cleaner, thanks!